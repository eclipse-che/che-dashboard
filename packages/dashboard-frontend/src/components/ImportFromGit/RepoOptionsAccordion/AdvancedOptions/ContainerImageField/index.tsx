/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (definition: string | undefined) => void;
  containerImage: string | undefined;
};
export type State = {
  containerImage: string | undefined;
};

export class ContainerImageField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      containerImage: this.props.containerImage,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { containerImage } = this.props;
    if (prevProps.containerImage !== containerImage) {
      this.setState({ containerImage });
    }
  }

  private handleChange(value: string) {
    let containerImage: string | undefined = value.trim();
    containerImage = containerImage !== '' ? containerImage : undefined;
    if (containerImage !== this.state.containerImage) {
      this.setState({ containerImage });
      this.props.onChange(containerImage);
    }
  }

  public render() {
    const containerImage = this.state.containerImage || '';

    return (
      <FormGroup label="Container Image">
        <TextInput
          aria-label="Container Image"
          placeholder="Enter the container image"
          onChange={value => this.handleChange(value)}
          value={containerImage}
        />
      </FormGroup>
    );
  }
}
