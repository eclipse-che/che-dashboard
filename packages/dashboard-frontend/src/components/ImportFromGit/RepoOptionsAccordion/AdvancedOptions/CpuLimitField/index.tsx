/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, Slider } from '@patternfly/react-core';
import React from 'react';

const steps = [
  { value: 0, label: 'default' },
  { value: 1, label: '1' },
  { value: 2, label: '2', isLabelHidden: true },
  { value: 3, label: '3', isLabelHidden: true },
  { value: 4, label: '4' },
  { value: 5, label: '5', isLabelHidden: true },
  { value: 6, label: '6', isLabelHidden: true },
  { value: 7, label: '7', isLabelHidden: true },
  { value: 8, label: '8' },
];

export type Props = {
  onChange: (cpuLimit: number) => void;
  cpuLimit: number;
};
export type State = {
  cpuLimit: number;
};

export class CpuLimitField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      cpuLimit: this.props.cpuLimit,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { cpuLimit } = this.props;
    if (prevProps.cpuLimit !== cpuLimit && cpuLimit !== this.state.cpuLimit) {
      this.setState({ cpuLimit });
    }
  }

  private handleChange(cpuLimit: number) {
    if (cpuLimit !== this.state.cpuLimit) {
      this.setState({ cpuLimit });
      this.props.onChange(cpuLimit);
    }
  }

  private getLabel(cpuLimit: number): string {
    const label = 'CPU Limit';
    if (cpuLimit === 0) {
      return label;
    } else if (cpuLimit === 1) {
      return `${label} (1 core)`;
    }

    return `${label} (${cpuLimit} cores)`;
  }

  public render() {
    const cpuLimit = this.state.cpuLimit;
    const label = this.getLabel(cpuLimit);

    return (
      <FormGroup label={label}>
        <Slider
          data-testid="cpu-limit-slider"
          value={cpuLimit}
          onChange={value => this.handleChange(value)}
          max={steps[steps.length - 1].value}
          customSteps={steps}
        />
      </FormGroup>
    );
  }
}
