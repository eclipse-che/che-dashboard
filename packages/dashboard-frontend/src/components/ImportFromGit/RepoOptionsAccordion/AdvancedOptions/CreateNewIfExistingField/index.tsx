/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, Switch } from '@patternfly/react-core';
import React from 'react';

import { Navigation } from '@/Layout/Navigation';
import { CREATE_NEW_IF_EXIST_SWITCH_ID } from '@/pages/GetStarted/SamplesList/Toolbar/CreateNewIfExistSwitch';

export type Props = {
  onChange: (createNewIfExisting: boolean | undefined) => void;
  createNewIfExisting: boolean | undefined;
  isHidden?: boolean;
};
export type State = {
  createNewIfExisting: boolean;
};

export class CreateNewIfExistingField extends React.PureComponent<Props, State> {
  private readonly subscribeCallback: (isChecked: boolean | undefined) => void;
  constructor(props: Props) {
    super(props);

    this.state = {
      createNewIfExisting: this.props.createNewIfExisting || false,
    };

    this.subscribeCallback = (isChecked: boolean | undefined) =>
      this.handleChange(isChecked === true);
  }

  public componentDidMount() {
    Navigation.pageState[CREATE_NEW_IF_EXIST_SWITCH_ID] = {
      isChecked: this.state.createNewIfExisting,
    };
    // Subscribe to the page state when the component is mounted
    Navigation.pageState.subscribe(this.subscribeCallback, CREATE_NEW_IF_EXIST_SWITCH_ID);
  }

  public componentWillUnmount() {
    // Unsubscribe from the page state when the component is unmounted
    Navigation.pageState.unsubscribe(this.subscribeCallback, CREATE_NEW_IF_EXIST_SWITCH_ID);
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const createNewIfExisting = this.props.createNewIfExisting || false;
    if (prevProps.createNewIfExisting !== createNewIfExisting) {
      this.setState({ createNewIfExisting });
    }
  }

  private handleChange(createNewIfExisting: boolean): void {
    if (createNewIfExisting === this.state.createNewIfExisting) {
      return; // No change, do nothing
    }
    this.setState({ createNewIfExisting });
    this.props.onChange(createNewIfExisting);
  }

  public render() {
    const { createNewIfExisting } = this.state;
    const display = this.props.isHidden ? 'none' : 'block';

    return (
      <FormGroup label="Create New If Existing" style={{ display }}>
        <Switch
          id="advanced-options-create-new-if-exist-switch"
          aria-label="Create New If Existing"
          isChecked={createNewIfExisting}
          onChange={value => this.handleChange(value)}
        />
      </FormGroup>
    );
  }
}
