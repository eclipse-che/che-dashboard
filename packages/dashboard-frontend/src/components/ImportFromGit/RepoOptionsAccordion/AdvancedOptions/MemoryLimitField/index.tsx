/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, Slider } from '@patternfly/react-core';
import React from 'react';

import { formatBytes } from '@/components/ImportFromGit/helpers';

export const STEP = 1073741824;

const steps = [
  { value: 0, label: 'default' },
  { value: 1, label: '1', isLabelHidden: true },
  { value: 2, label: '2', isLabelHidden: true },
  { value: 3, label: '4', isLabelHidden: true },
  { value: 4, label: '4' },
  { value: 6, label: '6', isLabelHidden: true },
  { value: 8, label: '8', isLabelHidden: true },
  { value: 12, label: '12', isLabelHidden: true },
  { value: 16, label: '16' },
  { value: 20, label: '20', isLabelHidden: true },
  { value: 24, label: '24', isLabelHidden: true },
  { value: 28, label: '28', isLabelHidden: true },
  { value: 32, label: '32' },
];

export type Props = {
  onChange: (memoryLimit: number) => void;
  memoryLimit: number;
};
export type State = {
  memoryLimit: number;
};

export class MemoryLimitField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    const memoryLimit = this.getMemoryLimit();

    this.state = {
      memoryLimit,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    if (prevProps.memoryLimit !== this.props.memoryLimit) {
      const memoryLimit = this.getMemoryLimit();
      if (memoryLimit !== this.state.memoryLimit) {
        this.setState({ memoryLimit });
      }
    }
  }

  private handleChange(memoryLimit: number) {
    if (memoryLimit !== this.state.memoryLimit) {
      this.setState({ memoryLimit });
      this.props.onChange(memoryLimit * STEP);
    }
  }

  private getMemoryLimit(): number {
    const memoryLimit = this.props.memoryLimit;
    if (memoryLimit <= STEP) {
      return 0;
    }
    return memoryLimit / STEP;
  }

  private getLabel(memoryLimit: number): string {
    if (memoryLimit > 0) {
      return `Memory Limit (${formatBytes(memoryLimit * STEP)})`;
    }

    return 'Memory Limit';
  }

  public render() {
    const memoryLimit = this.state.memoryLimit;
    const label = this.getLabel(memoryLimit);

    return (
      <FormGroup label={label}>
        <Slider
          data-testid="memory-limit-slider"
          value={memoryLimit}
          onChange={value => this.handleChange(value)}
          max={steps[steps.length - 1].value}
          customSteps={steps}
        />
      </FormGroup>
    );
  }
}
