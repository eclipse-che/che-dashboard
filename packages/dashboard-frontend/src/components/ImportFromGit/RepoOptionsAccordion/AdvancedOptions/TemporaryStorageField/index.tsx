/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, Switch } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (isTemporary: boolean | undefined) => void;
  isTemporary: boolean | undefined;
};
export type State = {
  isTemporary: boolean;
};

export class TemporaryStorageField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isTemporary: this.props.isTemporary || false,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const isTemporary = this.props.isTemporary || false;
    if (prevProps.isTemporary !== isTemporary) {
      this.setState({ isTemporary });
    }
  }

  private handleChange(isTemporary: boolean) {
    this.setState({ isTemporary });
    this.props.onChange(isTemporary);
  }

  public render() {
    const { isTemporary } = this.state;

    return (
      <FormGroup label="Temporary Storage">
        <Switch
          id="temporary-storage-switch"
          aria-label="Temporary Storage"
          isChecked={isTemporary}
          onChange={value => this.handleChange(value)}
        />
      </FormGroup>
    );
  }
}
