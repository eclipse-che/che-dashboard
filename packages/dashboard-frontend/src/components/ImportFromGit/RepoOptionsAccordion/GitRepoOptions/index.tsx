/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import { isEqual } from 'lodash';
import React from 'react';

import { AdditionalGitRemotes } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/AdditionalGitRemotes';
import { GitBranchField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/GitBranchField';
import { PathToDevfileField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/PathToDevfileField';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';

export type Props = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  hasSupportedGitService: boolean;
  onChange: (
    gitBranch: string | undefined,
    remotes: GitRemote[] | undefined,
    devfilePath: string | undefined,
    isValid: boolean,
  ) => void;
};

export type State = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  isValid: boolean;
};

export class GitRepoOptions extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitBranch: props.gitBranch,
      remotes: props.remotes,
      devfilePath: props.devfilePath,
      isValid: true,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { gitBranch, remotes, devfilePath } = this.props;

    if (gitBranch !== prevProps.gitBranch) {
      this.setState({ gitBranch });
    }

    if (!isEqual(remotes, prevProps.remotes)) {
      this.setState({ remotes });
    }

    if (devfilePath !== prevProps.devfilePath) {
      this.setState({ devfilePath });
    }
  }
  private handleGitBranch(gitBranch: string | undefined) {
    const { remotes, devfilePath, isValid } = this.state;

    this.setState({ gitBranch });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleRemotes(remotes: GitRemote[] | undefined, isValid: boolean) {
    const { gitBranch, devfilePath } = this.state;

    this.setState({ remotes, isValid });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleDevfilePath(devfilePath: string | undefined) {
    const { gitBranch, remotes, isValid } = this.state;

    this.setState({ devfilePath });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }
  public render() {
    const { hasSupportedGitService } = this.props;
    const { gitBranch, remotes, devfilePath } = this.state;
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        {hasSupportedGitService && (
          <GitBranchField
            onChange={gitBranch => this.handleGitBranch(gitBranch)}
            gitBranch={gitBranch}
          />
        )}
        <AdditionalGitRemotes
          onChange={(remotes: GitRemote[] | undefined, isValid: boolean) =>
            this.handleRemotes(remotes, isValid)
          }
          remotes={remotes}
        />
        <PathToDevfileField
          onChange={devfilePath => this.handleDevfilePath(devfilePath)}
          devfilePath={devfilePath}
        />
      </Form>
    );
  }
}
