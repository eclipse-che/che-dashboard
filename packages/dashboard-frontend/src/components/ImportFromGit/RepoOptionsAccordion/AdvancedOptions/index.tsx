/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import React from 'react';

import { ContainerImageField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/ContainerImageField';
import { CpuLimitField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CpuLimitField';
import { CreateNewIfExistingField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CreateNewIfExistingField';
import { MemoryLimitField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/MemoryLimitField';
import { TemporaryStorageField } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/TemporaryStorageField';

export type Props = {
  containerImage: string | undefined;
  temporaryStorage: boolean | undefined;
  createNewIfExisting: boolean | undefined;
  memoryLimit: number | undefined;
  cpuLimit: number | undefined;
  onChange: (
    containerImage: string | undefined,
    temporaryStorage: boolean | undefined,
    createNewIfExisting: boolean | undefined,
    memoryLimit: number | undefined,
    cpuLimit: number | undefined,
  ) => void;
};

export type State = {
  containerImage: string | undefined;
  temporaryStorage: boolean | undefined;
  createNewIfExisting: boolean | undefined;
  memoryLimit: number | undefined;
  cpuLimit: number | undefined;
};

export class AdvancedOptions extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      containerImage: props.containerImage,
      temporaryStorage: props.temporaryStorage,
      createNewIfExisting: props.createNewIfExisting,
      memoryLimit: props.memoryLimit,
      cpuLimit: props.cpuLimit,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } =
      this.props;

    if (containerImage !== prevProps.containerImage) {
      this.setState({ containerImage });
    }

    if (temporaryStorage !== prevProps.temporaryStorage) {
      this.setState({ temporaryStorage });
    }

    if (createNewIfExisting !== prevProps.createNewIfExisting) {
      this.setState({ createNewIfExisting });
    }

    if (memoryLimit !== prevProps.memoryLimit) {
      this.setState({ memoryLimit });
    }

    if (cpuLimit !== prevProps.cpuLimit) {
      this.setState({ cpuLimit });
    }
  }

  private handleContainerImage(containerImage: string | undefined) {
    const { temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } = this.state;

    this.setState({ containerImage });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleTemporaryStorage(temporaryStorage: boolean | undefined) {
    const { containerImage, createNewIfExisting, memoryLimit, cpuLimit } = this.state;

    this.setState({ temporaryStorage });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleCreateNewIfExisting(createNewIfExisting: boolean | undefined) {
    const { containerImage, temporaryStorage, memoryLimit, cpuLimit } = this.state;

    this.setState({ createNewIfExisting });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleMemoryLimit(memoryLimit: number | undefined) {
    const { containerImage, temporaryStorage, createNewIfExisting, cpuLimit } = this.state;

    this.setState({ memoryLimit });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  private handleCpuLimit(cpuLimit: number | undefined) {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit } = this.state;

    this.setState({ cpuLimit });
    this.props.onChange(
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
    );
  }

  public render() {
    const { containerImage, temporaryStorage, createNewIfExisting, memoryLimit, cpuLimit } =
      this.state;
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        <ContainerImageField
          onChange={containerImage => this.handleContainerImage(containerImage)}
          containerImage={containerImage}
        />
        <TemporaryStorageField
          onChange={temporaryStorage => this.handleTemporaryStorage(temporaryStorage)}
          isTemporary={temporaryStorage}
        />
        <CreateNewIfExistingField
          onChange={createNewIfExisting => this.handleCreateNewIfExisting(createNewIfExisting)}
          createNewIfExisting={createNewIfExisting}
        />
        <MemoryLimitField
          onChange={memoryLimit => this.handleMemoryLimit(memoryLimit)}
          memoryLimit={memoryLimit || 0}
        />
        <CpuLimitField
          onChange={cpuLimit => this.handleCpuLimit(cpuLimit)}
          cpuLimit={cpuLimit || 0}
        />
      </Form>
    );
  }
}
