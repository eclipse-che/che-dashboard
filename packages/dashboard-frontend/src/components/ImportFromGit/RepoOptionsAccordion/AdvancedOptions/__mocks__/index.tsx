/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions';

export class AdvancedOptions extends React.PureComponent<Props> {
  public render() {
    const {
      containerImage,
      temporaryStorage,
      createNewIfExisting,
      memoryLimit,
      cpuLimit,
      onChange,
    } = this.props;

    return (
      <div>
        <div>Advanced Options</div>
        <div data-testid="advanced-options">{`${containerImage}, ${temporaryStorage}, ${createNewIfExisting}, ${memoryLimit}, ${cpuLimit}`}</div>
        <button
          onClick={() =>
            onChange('newContainerImage', !temporaryStorage, !createNewIfExisting, 1073741824, 1)
          }
        >
          Advanced Options Change
        </button>
      </div>
    );
  }
}
