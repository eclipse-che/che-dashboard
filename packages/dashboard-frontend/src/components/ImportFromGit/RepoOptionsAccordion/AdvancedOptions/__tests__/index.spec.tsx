/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';

import { AdvancedOptions } from '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/ContainerImageField');
jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CpuLimitField');
jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/MemoryLimitField');
jest.mock('@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/TemporaryStorageField');
jest.mock(
  '@/components/ImportFromGit/RepoOptionsAccordion/AdvancedOptions/CreateNewIfExistingField',
);

const mockOnChange = jest.fn();

describe('AdvancedOptions', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot with default values', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with all values', () => {
    const snapshot = createSnapshot('testimage', true, true, 4718592, 2);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('update Container Image', () => {
    renderComponent('testimage');

    const containerImage = screen.getByTestId('container-image');

    expect(containerImage).toHaveTextContent('testimage');

    const updateContainerImage = screen.getByRole('button', {
      name: 'Container Image Change',
    });

    userEvent.click(updateContainerImage);

    expect(mockOnChange).toHaveBeenCalledWith(
      'new-container-image',
      undefined,
      undefined,
      undefined,
      undefined,
    );
  });

  test('update Cpu Limit', () => {
    renderComponent(undefined, undefined, undefined, undefined, 8);

    const cpuLimit = screen.getByTestId('cpu-limit');

    expect(cpuLimit).toHaveTextContent('8');

    const updateCpuLimit = screen.getByRole('button', {
      name: 'Cpu Limit Change',
    });

    userEvent.click(updateCpuLimit);

    expect(mockOnChange).toHaveBeenCalledWith(undefined, undefined, undefined, undefined, 1);
  });

  test('update CreateNewIfExisting', () => {
    renderComponent(undefined, undefined, true);

    const createNewIfExisting = screen.getByTestId('create-new-if-existing');

    expect(createNewIfExisting).toHaveTextContent('true');

    const updateCreateNewIfExisting = screen.getByRole('button', {
      name: 'Create New If Existing Change',
    });

    userEvent.click(updateCreateNewIfExisting);

    expect(mockOnChange).toHaveBeenCalledWith(undefined, undefined, false, undefined, undefined);
  });

  test('update Memory Limit', () => {
    renderComponent(undefined, undefined, undefined, 4718592);

    const memoryLimit = screen.getByTestId('memory-limit');

    expect(memoryLimit).toHaveTextContent('4718592');

    const updateMemoryLimit = screen.getByRole('button', {
      name: 'Memory Limit Change',
    });

    userEvent.click(updateMemoryLimit);

    expect(mockOnChange).toHaveBeenCalledWith(
      undefined,
      undefined,
      undefined,
      1073741824,
      undefined,
    );
  });

  test('update Temporary Storage', () => {
    renderComponent(undefined, true);

    const temporaryStorage = screen.getByTestId('temporary-storage');

    expect(temporaryStorage).toHaveTextContent('true');

    const updateTemporaryStorage = screen.getByRole('button', {
      name: 'Temporary Storage Change',
    });

    userEvent.click(updateTemporaryStorage);

    expect(mockOnChange).toHaveBeenCalledWith(undefined, false, undefined, undefined, undefined);
  });
});

function getComponent(
  containerImage?: string | undefined,
  temporaryStorage?: boolean | undefined,
  createNewIfExisting?: boolean | undefined,
  memoryLimit?: number | undefined,
  cpuLimit?: number | undefined,
) {
  return (
    <AdvancedOptions
      containerImage={containerImage}
      temporaryStorage={temporaryStorage}
      createNewIfExisting={createNewIfExisting}
      memoryLimit={memoryLimit}
      cpuLimit={cpuLimit}
      onChange={mockOnChange}
    />
  );
}
