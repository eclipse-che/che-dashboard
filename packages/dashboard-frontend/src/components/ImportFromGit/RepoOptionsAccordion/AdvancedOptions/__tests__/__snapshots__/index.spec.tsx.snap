// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedOptions snapshot with all values 1`] = `
<form
  className="pf-c-form pf-m-horizontal"
  noValidate={true}
  onSubmit={[Function]}
>
  <div>
    <div>
      Container Image
    </div>
    <div
      data-testid="container-image"
    >
      testimage
    </div>
    <button
      onClick={[Function]}
    >
      Container Image Change
    </button>
  </div>
  <div>
    <div>
      Temporary Storage
    </div>
    <div
      data-testid="temporary-storage"
    >
      true
    </div>
    <button
      onClick={[Function]}
    >
      Temporary Storage Change
    </button>
  </div>
  <div>
    <div>
      Create New If Existing
    </div>
    <div
      data-testid="create-new-if-existing"
    >
      true
    </div>
    <button
      onClick={[Function]}
    >
      Create New If Existing Change
    </button>
  </div>
  <div>
    <div>
      Memory Limit
    </div>
    <div
      data-testid="memory-limit"
    >
      4718592
    </div>
    <button
      onClick={[Function]}
    >
      Memory Limit Change
    </button>
  </div>
  <div>
    <div>
      Cpu Limit
    </div>
    <div
      data-testid="cpu-limit"
    >
      2
    </div>
    <button
      onClick={[Function]}
    >
      Cpu Limit Change
    </button>
  </div>
</form>
`;

exports[`AdvancedOptions snapshot with default values 1`] = `
<form
  className="pf-c-form pf-m-horizontal"
  noValidate={true}
  onSubmit={[Function]}
>
  <div>
    <div>
      Container Image
    </div>
    <div
      data-testid="container-image"
    />
    <button
      onClick={[Function]}
    >
      Container Image Change
    </button>
  </div>
  <div>
    <div>
      Temporary Storage
    </div>
    <div
      data-testid="temporary-storage"
    >
      undefined
    </div>
    <button
      onClick={[Function]}
    >
      Temporary Storage Change
    </button>
  </div>
  <div>
    <div>
      Create New If Existing
    </div>
    <div
      data-testid="create-new-if-existing"
    >
      undefined
    </div>
    <button
      onClick={[Function]}
    >
      Create New If Existing Change
    </button>
  </div>
  <div>
    <div>
      Memory Limit
    </div>
    <div
      data-testid="memory-limit"
    >
      0
    </div>
    <button
      onClick={[Function]}
    >
      Memory Limit Change
    </button>
  </div>
  <div>
    <div>
      Cpu Limit
    </div>
    <div
      data-testid="cpu-limit"
    >
      0
    </div>
    <button
      onClick={[Function]}
    >
      Cpu Limit Change
    </button>
  </div>
</form>
`;
